### PR TITLE
Add Vehicle component classes

### PR DESCRIPTION
**Pull request by kagikn**

### CODE REVIEW

File: source/scripting/Vehicle.cs [hunk 1]
```diff
@@ -361,6 +283,10 @@ public sealed class Vehicle : Entity
 	{
 		public Vehicle(int handle) : base(handle)
 		{
+			Doors = new VehicleDoorCollection(this);
+			Mods = new VehicleModCollection(this);
+			Wheels = new VehicleWheelCollection(this);
+			Windows = new VehicleWindowCollection(this);
```
Comment (zorg93) [inline source/scripting/Vehicle.cs]: Not a fan on initializing all these in the constructor, most of the time someone creates a vehicle class they aren't interested in the doors/mods/wheels/windows. better to initialize them in their respective getters. Also with how the INativeValue is currently set up, the constructor for Vehicle wont be called if its being returned from a native so calling the getters would result in a null reference error

**Comment by zorg93** [comment]
Would an enum for the wheels not be a nicer implementation, or do you not know the names of all possible wheels?

File: source/scripting/Vehicle.cs [hunk 2]
```diff
@@ -1627,78 +1557,38 @@ public bool CanWheelsBreak
...
-		}
 
 		public VehicleDoor[] GetDoors()
 		{
 			var list = new List<VehicleDoor>();
 
 			if (HasBone("door_dside_f"))
 			{
-				list.Add(VehicleDoor.FrontLeftDoor);
+				new VehicleDoor(this, VehicleDoorIndex.FrontLeftDoor);
```
Comment (crosire) [inline source/scripting/Vehicle.cs]: Might want to add the result to `list` =)

**Comment by kagikn** [comment]
I'm looking for how to convert a wheel ID to the wheel pointer(like RPH does).

**Comment by kagikn** [comment]
I couldn't find way to convert a wheel ID to the wheel pointer after all.
Jet(0x3F119114) has wheel ID 49 and 51, but these seem to share wheel the bones with the other wheel IDs. I couldn't find the way to detect what wheel IDs are assigned to what wheel pointers, so I removed the features that use pointer from ```VehicleWheel```. I'm so jealous of MulleDK19's docompiling ability!

File: source/scripting/VehicleDoor.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+		{
+			Function.Call(Hash.SET_VEHICLE_DOOR_OPEN, _owner.Handle, (int)Index, loose, instantly);
+		}
+		public void Close(bool instantly)
+		{
+			Function.Call(Hash.SET_VEHICLE_DOOR_SHUT, _owner.Handle, (int)Index, instantly);
+		}
+		public void BreakOff()
+		{
+			Function.Call(Hash.SET_VEHICLE_DOOR_BROKEN, _owner.Handle, (int)Index);
```
Comment (zorg93) [inline source/scripting/VehicleDoor.cs]: This is probably my fault, as i didnt include it in my commit, but there should be a 3rd parameter in that native, and it needs setting to true

Comment (kagikn) [inline source/scripting/VehicleDoor.cs]: I confirmed the 3rd parameter is the value if a door remains in the world or not.

File: source/scripting/VehicleDoor.cs [hunk 4]
```diff
@@ -75,7 +75,11 @@ public void Close(bool instantly)
 		}
 		public void BreakOff()
 		{
-			Function.Call(Hash.SET_VEHICLE_DOOR_BROKEN, _owner.Handle, (int)Index);
+			Function.Call(Hash.SET_VEHICLE_DOOR_BROKEN, _owner.Handle, (int)Index, true);
+		}
+		public void Delete()
+		{
+			Function.Call(Hash.SET_VEHICLE_DOOR_BROKEN, _owner.Handle, (int)Index, false);
```
Comment (zorg93) [inline source/scripting/VehicleDoor.cs]: If the 3rd param is whether the door stays in the world, the true and false should be switched

File: source/scripting/VehicleDoor.cs [hunk 5]
```diff
@@ -75,11 +75,11 @@ public void Close(bool instantly)
 		}
 		public void BreakOff()
 		{
-			Function.Call(Hash.SET_VEHICLE_DOOR_BROKEN, _owner.Handle, (int)Index, true);
+			BreakOff(true);
 		}
-		public void Delete()
+		public void BreakOff(bool stayInTheWorld)
```
Comment (crosire) [inline source/scripting/VehicleDoor.cs]: You may use default parameters instead here:
```
public void BreakOff(bool stayInTheWorld = true)
```

By the way, I'd prefer a `CanBeBroken` and `Break()` over the current names.

File: source/scripting/VehicleDoor.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+			Index = index;
+		}
+
+		public VehicleDoorIndex Index { get; private set; }
+
+		public float AngleRatio
+		{
+			get
+			{
+				return Function.Call<float>(Hash.GET_VEHICLE_DOOR_ANGLE_RATIO, _owner.Handle, (int)Index);
```
Comment (crosire) [inline source/scripting/VehicleDoor.cs]: You don't need the cast to int. `GTA.Native.Function.Call()` accepts enums now.

File: source/scripting/VehicleDoor.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+			}
+		}
+		public Vehicle Vehicle
+		{
+			get { return _owner; }
+		}
+
+		public void Open(bool instantly)
+		{
+			Open(instantly);
```
Comment (JohnnyCrazy) [inline source/scripting/VehicleDoor.cs]: This...can't be right? :sweat_smile: 

Maybe also consider reordering parameters, or like @crosire said, default parameters.  
`public void Open(bool instantly, bool loose = false)`
would be nice IMO

File: source/scripting/VehicleDoorCollection.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        BackLeftDoor = 2,
+        Hood = 4,
+        Trunk
+    }
+
+    public sealed class VehicleDoorCollection
+    {
+        #region Fields
+        Vehicle _owner;
+        readonly Dictionary<int, VehicleDoor> _vehicleDoors = new Dictionary<int, VehicleDoor>();
```
Comment (crosire) [inline source/scripting/VehicleDoorCollection.cs]: Why not `Dictionary<VehicleDoorIndex, VehicleDoor>`? Gets rid of the cast to int below that way.

**Comment by JohnnyCrazy** [comment]
Very nice :+1:

File: source/scripting/VehicleDoorCollection.cs (hunk 8 above)
Comment (kagikn) [inline source/scripting/VehicleDoorCollection.cs]: If we use enums as dictionary keys without ```IEqualityComparer```, boxing problem occurs.

**Pull request merged**